compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3729,6 +3742,13 @@ private BType createArrayAndTupleIntersection(BArrayType arrayType, BTupleType t
             tupleMemberTypes.add(intersectionType);
         }
 
+        if (tupleType.restType == null) {
+            if (arrayType.size == tupleType.tupleTypes.size() || arrayType.state == BArrayState.OPEN) {
+                return new BTupleType(null, tupleMemberTypes);
+            }
+            return symTable.semanticError;
+        }
```

Comment:
Shouldn't the size-based checks be done before the for loop checking for individual type intersections? If the size doesn't match there's no intersection anyway and this is the less expensive check?